lib/srv/db/dynamodb/engine.go

```diff
@@ -0,0 +1,373 @@
+	// required by AWS for request validation: https://docs.aws.amazon.com/general/latest/gr/signing-elements.html
+	signingSvc *libaws.SigningService
+	// EngineConfig is the common database engine configuration.
+	common.EngineConfig
+	// clientConn is a client connection.
+	clientConn net.Conn
+	// sessionCtx is current session context.
+	sessionCtx *common.Session
+	// RoundTrippers is a cache of RoundTrippers, mapped by service endpoint.
+	// It is not guarded by a mutex, since requests are processed serially.
+	RoundTrippers map[string]http.RoundTripper
+	// GetSigningCredsFn allows to set the function responsible for obtaining STS credentials.
+	// Used in tests to set static AWS credentials and skip API call.
+	GetSigningCredsFn libaws.GetSigningCredentialsFunc
+}
+
+var _ common.Engine = (*Engine)(nil)
+
+func (e *Engine) InitializeConnection(clientConn net.Conn, sessionCtx *common.Session) error {
+	e.clientConn = clientConn
+	e.sessionCtx = sessionCtx
+	awsSession, err := e.CloudClients.GetAWSSession(sessionCtx.Database.GetAWS().Region)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	svc, err := libaws.NewSigningService(libaws.SigningServiceConfig{
+		Clock:                 e.Clock,
+		Session:               awsSession,
+		GetSigningCredentials: e.GetSigningCredsFn,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	e.signingSvc = svc
+	return nil
+}
+
+// jsonErr is used to marshal a JSON error response as the AWS cli expects for errors.
+type jsonErr struct {
+	Code    string `json:"__type"`
+	Message string `json:"message"`
+}
+
+// SendError sends an error to DynamoDB client.
+func (e *Engine) SendError(err error) {
+	if e.clientConn == nil || err == nil || utils.IsOKNetworkError(err) {
+		return
+	}
+	e.Log.Debugf("DynamoDB connection error: %v", err)
+
+	// try to convert to a trace err if we can.
+	code := trace.ErrorToCode(err)
+	body, err := json.Marshal(jsonErr{
+		Code:    strconv.Itoa(code),
+		Message: err.Error(),
+	})
+	if err != nil {
+		e.Log.WithError(err).Error("failed to marshal error response")
+		return
+	}
+	response := &http.Response{
+		Status:     http.StatusText(code),
+		StatusCode: code,
+		Proto:      "HTTP/1.1",
+		ProtoMajor: 1,
+		ProtoMinor: 1,
+		// ContentLength is the authoritative value in the response,
+		// no need to also add the "Content-Length" header (source: go doc http.Response.Header).
+		ContentLength: int64(len(body)),
+		Header: map[string][]string{
+			"Content-Type": {"application/json"},
+		},
+		Body: ioutil.NopCloser(bytes.NewBuffer(body)),
+	}
+
+	if err := response.Write(e.clientConn); err != nil {
+		e.Log.WithError(err).Errorf("failed to send error response to DynamoDB client")
+		return
+	}
+}
+
+// HandleConnection authorizes the incoming client connection, connects to the
+// target DynamoDB server and starts proxying requests between client/server.
+func (e *Engine) HandleConnection(ctx context.Context, _ *common.Session) (err error) {
```

Comment:
thanks for catching, I agree and generally prefer to avoid named return. These places are just remnants of me putting in `defer .. if err != nil` debug statements that I missed when cleaning up my changes